data/src/main/java/com/microsoft/azure/kusto/data/auth/PublicAppTokenProviderBase.java

```diff
@@ -0,0 +1,31 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package com.microsoft.azure.kusto.data.auth;
+
+import com.microsoft.aad.msal4j.*;
+import org.jetbrains.annotations.NotNull;
+
+import java.net.MalformedURLException;
+import java.net.URISyntaxException;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+public abstract class PublicAppTokenProviderBase extends TokenProviderBase {
+    // TODO: Get ClientId from CM endpoint
+    static final String CLIENT_ID = "db662dc1-0cfe-4e1c-a843-19a68e65be58";
+    IPublicClientApplication clientApplication;
+
+    PublicAppTokenProviderBase(@NotNull String clusterUrl, String authorityId) throws URISyntaxException {
+        super(clusterUrl, authorityId);
+    }
+
+    @Override
+    protected IAuthenticationResult acquireAccessTokenSilentlyMsal() throws MalformedURLException, InterruptedException, ExecutionException, TimeoutException {
+        CompletableFuture<Set<IAccount>> accounts = clientApplication.getAccounts();
+        return clientApplication.acquireTokenSilently(getSilentParameters(accounts.join())).get(TIMEOUT_MS, TimeUnit.MILLISECONDS);

```

Comment:
I think you need to either user a user hint or make sure you only have one account only

Reply:
I'm not sure. I'll ask MSAL about this, since I couldn't discern proper behavior from their documentation.

I just looked for another example implementation to see what they do, and at https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java#L721 they take the first one for PublicApps.

Reply:
What would happen if I were to use a different identity when prompted for the second time?

Reply:
As a precaution, I reset the PublicClientApplication each time. There can only be one account/identity.